python/lsst/pipe/tasks/finalizeCharacterization.py

```diff
@@ -0,0 +1,693 @@
+
+        reserved_arr = np.zeros(len(selected_src), dtype=bool)
+        reserved_arr[matched_src] = isolated_source_table['reserved'][matched_iso]
+        selected_src['final_psf_reserved'] = reserved_arr
+
+        selected_src = selected_src[selected_src['final_psf_candidate']].copy(deep=True)
+
+        # Make the measured source catalog as well, based on the selected catalog.
+        measured_src = afwTable.SourceCatalog(self.schema)
+        measured_src.reserve(len(selected_src))
+        measured_src.extend(selected_src, mapper=self.schema_mapper)
+
+        # We need to copy over the final_psf flags because they were not in the mapper
+        measured_src['final_psf_candidate'] = selected_src['final_psf_candidate']
+        measured_src['final_psf_reserved'] = selected_src['final_psf_reserved']
+
+        # Select the psf candidates from the selection catalog
+        try:
+            psf_selection_result = self.make_psf_candidates.run(selected_src, exposure=exposure)
+        except Exception as e:
+            self.log.warn('Failed to make psf candidates for visit %d, detector %d: %s',
+                          visit, detector, e)
+            return None, None, measured_src
+
+        psf_cand_cat = psf_selection_result.goodStarCat
+
+        # Make list of psf candidates to send to the determiner
+        # (omitting those marked as reserved)
+        psf_determiner_list = [cand for cand, use
+                               in zip(psf_selection_result.psfCandidates,
+                                      ~psf_cand_cat['final_psf_reserved']) if use]
+        flag_key = psf_cand_cat.schema['final_psf_used'].asKey()
+        try:
+            psf, cell_set = self.psf_determiner.determinePsf(exposure,
+                                                             psf_determiner_list,
+                                                             self.metadata,
+                                                             flagKey=flag_key)
+        except Exception as e:
+            self.log.warn('Failed to determine psf for visit %d, detector %d: %s',
+                          visit, detector, e)
+            return None, None, measured_src
+
+        # At this point, we need to transfer the psf used flag from the selection
+        # catalog to the measurement catalog.
+        matched_selected, matched_measured = esutil.numpy_util.match(
+            selected_src['id'],
+            measured_src['id']
+        )
+        measured_used = np.zeros(len(measured_src), dtype=bool)
+        measured_used[matched_measured] = selected_src['final_psf_used'][matched_selected]
+        measured_src['final_psf_used'] = measured_used
+
+        # Next, we do the measurement on the psf stars ...
+        try:
+            self.measurement.run(measCat=measured_src, exposure=exposure)
+        except Exception as e:
+            self.log.warn('Failed to make measurements for visit %d, detector %d: %s',
```

Comment:
I think I've been told that we're supposed to say `warning` rather than `warn` now.